src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -51,59 +50,70 @@ class Serializer implements SerializerInterface, NormalizerInterface, Denormaliz
      */
     public const EMPTY_ARRAY_AS_OBJECT = 'empty_array_as_object';
 
-    private const SCALAR_TYPES = [
-        'int' => true,
-        'bool' => true,
-        'float' => true,
-        'string' => true,
-    ];
+    private NormalizerInterface $normalizer;
+
+    private DenormalizerInterface $denormalizer;
 
     protected ChainEncoder $encoder;
 
     protected ChainDecoder $decoder;
 
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $denormalizerCache = [];
-
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $normalizerCache = [];
-
     /**
      * @param array<NormalizerInterface|DenormalizerInterface> $normalizers
      * @param array<EncoderInterface|DecoderInterface>         $encoders
      */
     public function __construct(
-        private array $normalizers = [],
+        array $normalizers = [],
         array $encoders = [],
+        NormalizerInterface $normalizer = null,
+        DenormalizerInterface $denormalizer = null,
```

Comment:
```suggestion
        ?NormalizerInterface $normalizer = null,
        ?DenormalizerInterface $denormalizer = null,
```
To fit with the actual CS